{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "600bec54",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sc.settings.verbosity = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "34ef80c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify data, setting, and paths\n",
    "settings = {\"Xenium_5K_BC\": {\"x_shift\": 0, \"y_shift\": 7000},\n",
    "            \"Xenium_5K_OC\": {\"x_shift\": 12000, \"y_shift\": 10000},\n",
    "            \"Xenium_5K_CC\": {\"x_shift\": 26000, \"y_shift\": 8000},\n",
    "            \"Xenium_5K_LC\": {\"x_shift\": 12000, \"y_shift\": 0},\n",
    "            \"Xenium_5K_Prostate\": {\"x_shift\": 26000, \"y_shift\": 1000},\n",
    "            \"Xenium_5K_Skin\": {\"x_shift\": 0, \"y_shift\": 1000}}\n",
    "\n",
    "output_dir = \"../../output/merged_data/\"\n",
    "save_dir = \"../../data/merged_data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3be5db21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes: 5001\n",
      "35641064 out of 87146596 transcripts in tumor cells.\n",
      "Number of genes: 5001\n",
      "81798615 out of 129043380 transcripts in tumor cells.\n",
      "Number of genes: 5001\n",
      "43489382 out of 90482395 transcripts in tumor cells.\n",
      "Number of genes: 5001\n",
      "71378586 out of 176649914 transcripts in tumor cells.\n",
      "Number of genes: 5001\n",
      "32109068 out of 59566887 transcripts in tumor cells.\n",
      "Number of genes: 5001\n",
      "37837245 out of 73302888 transcripts in tumor cells.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1903003 × 5001\n",
       "    obs: 'cell_id', 'global_x', 'global_y', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'cell_type_merged', 'batch'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'gene'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Construct merged data\n",
    "transcripts_dict = {}\n",
    "adata_dict = {}\n",
    "\n",
    "for data in settings.keys():\n",
    "    \n",
    "    # Read data\n",
    "    data_dir = f\"../../data/{data}/\"\n",
    "    adata = sc.read_h5ad(data_dir + \"intermediate_data/adata.h5ad\")\n",
    "    print(f\"Number of genes: {adata.n_vars}\")\n",
    "    \n",
    "    # Delete uns\n",
    "    del adata.uns\n",
    "    \n",
    "    # Shift coordinates\n",
    "    x_shift = settings[data][\"x_shift\"]\n",
    "    y_shift = settings[data][\"y_shift\"]\n",
    "    adata.obs[\"global_x\"] = adata.obs[\"global_x\"] + x_shift\n",
    "    adata.obs[\"global_y\"] = adata.obs[\"global_y\"] + y_shift\n",
    "    \n",
    "    # Select tumor cells\n",
    "    adata_tumor = adata[adata.obs[\"cell_type_merged\"] == \"Malignant cell\"].copy()\n",
    "    \n",
    "    # Read transcripts and select those within tumor cells\n",
    "    transcripts = pd.read_parquet(data_dir + \"processed_data/transcripts.parquet\")\n",
    "    transcripts_tumor = transcripts[transcripts[\"cell_id\"].isin(adata_tumor.obs[\"cell_id\"])]\n",
    "    print(f\"{transcripts_tumor.shape[0]} out of {transcripts.shape[0]} transcripts in tumor cells.\")\n",
    "    \n",
    "    # Store data\n",
    "    adata_dict[data] = adata\n",
    "    transcripts_dict[data] = transcripts_tumor\n",
    "\n",
    "adata = anndata.concat(adata_dict, axis = 0, merge = \"same\", label = \"batch\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fd9cd9d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Overall in-cytoplasm ratio in tumor cells\n",
    "transcripts_all = pd.concat(transcripts_dict.values(), axis = 0)\n",
    "gene_means = transcripts_all.groupby(\"target\")[\"overlaps_nucleus\"].mean().reset_index()\n",
    "gene_means.columns = [\"gene\", \"in_nucleus_ratio\"]\n",
    "gene_means = gene_means.sort_values(by = \"in_nucleus_ratio\", ascending = True)\n",
    "gene_means[\"in_cytoplasm_ratio\"] = 1 - gene_means[\"in_nucleus_ratio\"]\n",
    "gene_means.to_csv(output_dir + \"in_cytoplasm_ratio_tumor.csv\", index = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "46ca8529",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plot size: (16, 10)\n"
     ]
    }
   ],
   "source": [
    "# Determine plot size\n",
    "x_range = adata.obs[\"global_x\"].max() - adata.obs[\"global_x\"].min()\n",
    "y_range = adata.obs[\"global_y\"].max() - adata.obs[\"global_y\"].min()\n",
    "short_edge = min(x_range, y_range)\n",
    "\n",
    "scale = 10 / short_edge\n",
    "plot_figsize = (int(x_range * scale), int(y_range * scale))\n",
    "print(f\"Plot size: {plot_figsize}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5c76ba9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot batch\n",
    "sc.set_figure_params(figsize = plot_figsize)\n",
    "ax = sc.pl.scatter(adata, x=\"global_x\", y=\"global_y\", color=\"batch\", size=0.5, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_visible(False)\n",
    "plt.savefig(output_dir + \"batch.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()\n",
    "\n",
    "# Plot cell type\n",
    "sc.set_figure_params(figsize = plot_figsize)\n",
    "ax = sc.pl.scatter(adata, x=\"global_x\", y=\"global_y\", color=\"cell_type_merged\", size=0.5, show=False)\n",
    "ax.grid(False)\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_title(\"\")\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_visible(False)\n",
    "plt.savefig(output_dir + \"cell_type_merged.jpeg\", dpi = 300, bbox_inches = \"tight\")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e24fb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad(save_dir + \"adata_all_raw.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6acb3102",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AIVC-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
